client/kv.go

```diff
@@ -101,84 +91,44 @@ func (kv *KV) Sender() KVSender {
+	if len(calls) == 0 {
+		return nil
 	}
-	if args.Header().UserPriority == nil && kv.UserPriority != 0 {
-		args.Header().UserPriority = gogoproto.Int32(kv.UserPriority)
-	}
-	call := &Call{
-		Method: method,
-		Args:   args,
-		Reply:  reply,
-	}
-	call.resetClientCmdID(kv.clock)
-	kv.sender.Send(call)
-	err := call.Reply.Header().GoError()
-	if err != nil {
-		log.Infof("failed %s: %s", call.Method, err)
-	}
-	return err
-}
 
-// Prepare accepts a KV API call, specified by method name, arguments
-// and a reply struct. The call will be buffered locally until the
-// first call to Flush(), at which time it will be sent for execution
-// as part of a batch call. Using Prepare/Flush parallelizes queries
-// and updates and should be used where possible for efficiency.
-//
-// For clients using an HTTP sender, Prepare/Flush allows multiple
-// commands to be sent over the same connection. For transactional
-// clients, Prepare/Flush can dramatically improve efficiency by
-// compressing multiple writes into a single atomic update in the
-// event that the writes are to keys within a single range. However,
-// using Prepare/Flush alone will not guarantee atomicity. Clients
-// must use a transaction for that purpose.
-//
-// The supplied reply struct will not be valid until after a call
-// to Flush().
-func (kv *KV) Prepare(method string, args proto.Request, reply proto.Response) {
-	call := &Call{
-		Method: method,
-		Args:   args,
-		Reply:  reply,
+	// First check if any call contains an error.
+	for _, call := range calls {
+		if call.Err != nil {
+			return call.Err
+		}
 	}
-	call.resetClientCmdID(kv.clock)
-	kv.prepared = append(kv.prepared, call)
-}
 
-// Flush sends all previously prepared calls, buffered by invocations
-// of Prepare(). The calls are organized into a single batch command
-// and sent together. Flush returns nil if all prepared calls are
-// executed successfully. Otherwise, Flush returns the first error,
-// where calls are executed in the order in which they were prepared.
-// After Flush returns, all prepared reply structs will be valid.
-func (kv *KV) Flush() (err error) {
-	if len(kv.prepared) == 0 {
-		return
-	} else if len(kv.prepared) == 1 {
-		call := kv.prepared[0]
-		kv.prepared = []*Call{}
-		err = kv.Call(call.Method, call.Args, call.Reply)
+	if len(calls) == 1 {
+		c := calls[0]
+		if c.Args.Header().User == "" {
+			c.Args.Header().User = kv.User
+		}
+		if c.Args.Header().UserPriority == nil && kv.UserPriority != 0 {
+			c.Args.Header().UserPriority = gogoproto.Int32(kv.UserPriority)
+		}
+		c.resetClientCmdID(kv.clock)
+		kv.sender.Send(c)
+		err = c.Reply.Header().GoError()
+		if err != nil {
+			log.Infof("failed %s: %s", c.Method, err)
+		}
 		return
 	}
-	replies := make([]proto.Response, 0, len(kv.prepared))
+
+	replies := make([]proto.Response, 0, len(calls))
 	bArgs, bReply := &proto.BatchRequest{}, &proto.BatchResponse{}
-	for _, call := range kv.prepared {
+	for _, call := range calls {
```

Comment:
Why aren't you setting user, user priority & client cmd id in the batch? If you notice, the old code in flush would invoke Call with the batch which in turn would set these values as appropriate.